### PR TITLE
heartbeat race conditions fixed

### PR DESCRIPTION
**Pull request by raduchis**

### CODE REVIEW

File: node/heartbeat/monitor.go [hunk 1]
```diff
@@ -169,21 +170,15 @@ func (m *Monitor) ProcessReceivedMessage(message p2p.MessageP2P) error {
...
 
 func (m *Monitor) addHeartbeatMessageToMap(hb *Heartbeat) {
-	m.mutHeartbeatMessages.Lock()
-	defer m.mutHeartbeatMessages.Unlock()
-
 	pubKeyStr := string(hb.Pubkey)
+	m.mutHeartbeatMessages.Lock()
 	hbmi, ok := m.heartbeatMessages[pubKeyStr]
 	if hbmi == nil || !ok {
 		var err error
```
Comment (sasurobert) [inline node/heartbeat/monitor.go]: add m.mutheartbeatmessages.unlock to line 188

File: node/heartbeat/monitor.go [hunk 2]
```diff
@@ -243,31 +245,33 @@ func (m *Monitor) computeShardID(pubkey string) uint32 {
 }
 
 func (m *Monitor) computeAllHeartbeatMessages() {
+	m.mutHeartbeatMessages.Lock()
```
Comment (iulianpascalau) [inline node/heartbeat/monitor.go]: I do not like nested mutex operations. They may lead to deadlocks. 
You can use updateMutex inside heartbeatMessageInfo instances (can move the lock/unlock operations there)

Comment (sasurobert) [reply node/heartbeat/monitor.go]: +1

File: node/heartbeat/monitor.go (hunk 1 above)
Comment (raduchis) [reply node/heartbeat/monitor.go]: fixed

File: node/heartbeat/monitor.go (hunk 2 above)
Comment (iulianpascalau) [reply node/heartbeat/monitor.go]: heartbeatMessageInfo.updateMutex should be used exclusively inside heartbeatMessageInfo struct. The monitor struct that uses heartbeatMessageInfo instances will need to know too much to protect some operations by a mutex.

Comment (raduchis) [reply node/heartbeat/monitor.go]: Fixed

**Pull request merged**